### PR TITLE
Hosting & other Servers improvements

### PR DESCRIPTION
**Pull request by benaadams**
Betterize Hosting Log https://github.com/aspnet/AspNetCore/pull/10102

Pool the HostingApplication.Context and improve allocations of other servers

HttpSys for 1M requests https://github.com/aspnet/AspNetCore/pull/10180#issuecomment-492859731

* Allocates 34M objects rather than 54M objects (-20M, -37%)
* Allocates 1.64GB rather than 7.83GB (-6.19GB, -79%)

/cc @halter73

### CODE REVIEW

**Comment by davidfowl** [comment]
Better for Kestrel, worse for all the servers that can't pool it...

**Comment by benaadams** [comment]
> Better for Kestrel, worse for all the servers that can't pool it... 

Might be able to improve the other servers at the same time... Investigating

**Comment by benaadams** [comment]
HttpSys was boxing the struct so allocating it anyway; have changed it not to box

**Comment by benaadams** [comment]
Also shaves 32 bytes off the size of the `ProcessRequests` statemachine
```diff
 ; Assembly listing for method HttpProtocol:ProcessRequests(ref):ref:this
 ;
 ;  V00 this         [V00,T00] (  3,  3   )     ref  ->  rcx         this class-hnd
 ;  V01 arg1         [V01,T01] (  3,  3   )     ref  ->  rdx         class-hnd
-;  V02 loc0         [V02    ] (  7,  6.25)  struct (136) [rsp+0x20]   do-not-enreg[XSFB] must-init addr-exposed ld-addr-op
+;  V02 loc0         [V02    ] (  7,  6.25)  struct (104) [rsp+0x20]   do-not-enreg[XSFB] must-init addr-exposed ld-addr-op
 ;* V03 loc1         [V03    ] (  0,  0   )  struct ( 8) zero-ref    ld-addr-op
 ;  V04 OutArgs      [V04    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
 ;* V05 tmp1         [V05    ] (  0,  0   )  struct ( 8) zero-ref    ld-addr-op "Inline ldloca(s) first use temp"
 ;  V06 tmp2         [V06,T02] (  2,  4   )     ref  ->  rax         class-hnd "dup spill"
 ;  V07 tmp3         [V07,T03] (  4,  3.25)     ref  ->  rax        
 ;* V08 tmp4         [V08    ] (  0,  0   )     ref  ->  zero-ref    V03.m_task(offs=0x00) P-INDEP "field V03.m_task (fldOffset=0x0)"
 ;* V09 tmp5         [V09    ] (  0,  0   )     ref  ->  zero-ref    V05.m_task(offs=0x00) P-INDEP "field V05.m_task (fldOffset=0x0)"
 ;
 ; Lcl frame size = 168
 
 G_M39292_IG01:
        57                   push     rdi
        56                   push     rsi
-       4881ECA8000000       sub      rsp, 168
+       4881EC88000000       sub      rsp, 136
        488BF1               mov      rsi, rcx
        488D7C2420           lea      rdi, [rsp+20H]
-       B922000000           mov      ecx, 34
+       B91A000000           mov      ecx, 26
        33C0                 xor      rax, rax
        F3AB                 rep stosd 
```
And saves a bunch of stack copies from it
image

and

image

File: src/Hosting/Hosting/src/Internal/HostingApplication.cs [hunk 1]
```diff
@@ -29,40 +30,134 @@ public class HostingApplication : IHttpApplication<HostingApplication.Context>
...
 
         // Set up the request
-        public Context CreateContext(IFeatureCollection contextFeatures)
+        public HostContext CreateContext(IFeatureCollection contextFeatures)
         {
-            var context = new Context();
             var httpContext = _httpContextFactory.Create(contextFeatures);
 
-            _diagnostics.BeginRequest(httpContext, ref context);
+            if (contextFeatures is IContextContainer<HostContext> containter &&
```
Comment (DAllanCarr) [inline src/Hosting/Hosting/src/Internal/HostingApplication.cs]: Typo I think...

Comment (benaadams) [reply src/Hosting/Hosting/src/Internal/HostingApplication.cs]: Fixd and rebased

**Comment by benaadams** [comment]
> Better for Kestrel, worse for all the servers that can't pool it...

Enabled pooling of both the host Context and the HttpContext for all servers

**Comment by benaadams** [comment]
Can do better

**Comment by benaadams** [comment]
> > Better for Kestrel, worse for all the servers that can't pool it...

> Might be able to improve the other servers at the same time... Investigating

HttpSys for 1M requests:
* Allocates 34M objects rather than 54M objects (-20M, -37%)
* Allocates 1.64GB rather than 7.83GB (-6.19GB, -79%)

Before
image

After
image


Looking at IIS

**Comment by benaadams** [comment]
Splitting this up... first is https://github.com/aspnet/AspNetCore/pull/10283

File: src/Hosting/Hosting/src/Internal/HostingApplication.cs [hunk 2]
```diff
@@ -29,40 +30,132 @@ public class HostingApplication : IHttpApplication<HostingApplication.Context>
...
             var httpContext = context.HttpContext;
+
             _diagnostics.RequestEnd(httpContext, exception, context);
             _httpContextFactory.Dispose(httpContext);
             _diagnostics.ContextDisposed(context);
+
+            context.Reset();
         }
 
+        // Struct to turn {TContext} uses into faster concrete generic implementations rather than shared generics.
```
Comment (halter73) [inline src/Hosting/Hosting/src/Internal/HostingApplication.cs]: Just saw this. 👍 Not only does it help future maintainers, it heads off inevitable code review feedback from me 😄

**Comment by benaadams** [comment]
Superceaded by https://github.com/aspnet/AspNetCore/pull/10319

**Pull request closed without merging**